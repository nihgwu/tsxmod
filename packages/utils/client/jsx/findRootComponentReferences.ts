import { Node, SyntaxKind } from 'ts-morph'
import { findClosestComponentDeclaration } from './findClosestComponentDeclaration'
import { findReferencesAsJsxElements } from './findReferencesAsJsxElements'

/** Traces component references to the root component. */
export function findRootComponentReferences(node: Node): Node[] {
  const findReferences = (_node: Node, _allNodes: Node[] = []): Node[] => {
    const closestComponentIdentifier = findClosestComponentDeclaration(
      _node
    )?.getFirstDescendantByKind(SyntaxKind.Identifier)

    if (!closestComponentIdentifier) {
      return _allNodes
    }

    const elementReferences = findReferencesAsJsxElements(
      closestComponentIdentifier
    )

    if (elementReferences.length > 0) {
      return elementReferences.flatMap((reference) =>
        findReferences(reference, _allNodes)
      )
    }

    _allNodes.push(closestComponentIdentifier)

    return _allNodes
  }
  const allReferences = findReferences(node)

  return Array.from(new Set(allReferences))
}
