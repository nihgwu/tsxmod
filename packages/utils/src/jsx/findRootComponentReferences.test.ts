import { Project } from 'ts-morph'
import { findRootComponentReferences } from './findRootComponentReferences'

const buttonSource = `
import React from 'react'

export const Button = (props: {
    onClick: () => void
    children: React.ReactNode
}) => {
    return <button onClick={props.onClick}>{props.children}</button>
}
`

const dialogSource = `
import React from 'react'
import { Button } from './Button'

export function Dialog(props: {
    onClose: () => void
    children: React.ReactNode
}) {
    return (
        <dialog>
            {props.children}
            <Button onClick={props.onClose}>Close</Button>
        </dialog>
    )
}
`

const pageSource = `
import React from 'react'
import { Dialog } from './Dialog'

export default function Page() {
    return <div><Dialog /></div>
}
`

test('traces component references to the root element', () => {
  const project = new Project({ useInMemoryFileSystem: true })

  const buttonSourceFile = project.createSourceFile('Button.tsx', buttonSource)

  project.createSourceFile('Dialog.tsx', dialogSource)
  project.createSourceFile('about.tsx', pageSource)
  project.createSourceFile('contact.tsx', pageSource)

  const component = buttonSourceFile.getVariableDeclarationOrThrow('Button')
  const references = findRootComponentReferences(component)

  expect(
    references.map((node) => node.getSourceFile().getFilePath())
  ).toMatchSnapshot()
})
