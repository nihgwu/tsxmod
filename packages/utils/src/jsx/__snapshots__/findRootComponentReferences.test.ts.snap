// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`traces component references to the root element 1`] = `
[
  "/about.tsx",
  "/contact.tsx",
]
`;
